conbench/api/_google.py

```diff
@@ -1,44 +1,72 @@
+log = logging.getLogger(__name__)
+
+# Rely on INTENDED_BASE_URL to have a trailing slash.
+OIDC_CALLBACK_URL = Config.INTENDED_BASE_URL + "api/google/callback"
+
+
+def get_oidc_config():
     client_id = os.environ.get("GOOGLE_CLIENT_ID", None)
     client_secret = os.environ.get("GOOGLE_CLIENT_SECRET", None)
-    discovery_url = "https://accounts.google.com/.well-known/openid-configuration"
+    discovery_url = Config.OIDC_ISSUER_URL + "/.well-known/openid-configuration"
     return discovery_url, client_id, client_secret
 
 
-def get_google_client():
+def get_oidc_client():
     from oauthlib.oauth2 import WebApplicationClient
 
-    discovery_url, client_id, _ = get_google_config()
-    google = requests.get(discovery_url).json()
+    discovery_url, client_id, _ = get_oidc_config()
+
+    # Pragmatic healing for transient errors. Better: cache OP config across
+    # login requests. Also, if we want to retry here in the future: consider
+    # using tenacity.
+    for attempt in range(4):
+        try:
+            oidc_provider_config = requests.get(discovery_url).json()
+            break
+        except requests.exceptions.RequestException as exc:
+            log.info("err getting OP config (attempt %s): %s -- retry", attempt, exc)
+            time.sleep(2)
+
     client = WebApplicationClient(client_id)
-    return client, google
+    return client, oidc_provider_config
 
 
 def auth_google_user():
-    client, google = get_google_client()
-    redirect_uri = f.url_for("api.callback", _external=True, _scheme="https")
+    client, oidc_provider_config = get_oidc_client()
+
+    # http://127.0.0.1:5000/api/google/callback
+    # redirect_uri = f.url_for("api.callback", _external=True, _scheme="https")
+
+    # If either redirect URL or the authorization endpoint (at the OP) do not
+    # use the HTTPS scheme then the function below is expected to throw
+    # `InsecureTransportError`. For testing, this can be changed by setting
+    # the environment variable OAUTHLIB_INSECURE_TRANSPORT.
+
     return client.prepare_request_uri(
-        google["authorization_endpoint"],
-        redirect_uri=redirect_uri,
+        oidc_provider_config["authorization_endpoint"],
+        redirect_uri=OIDC_CALLBACK_URL,
         scope=["openid", "email", "profile"],
     )
 
 
 def get_google_user():
-    client, google = get_google_client()
-    _, client_id, client_secret = get_google_config()
+    client, oidc_provider_config = get_oidc_client()
+    _, client_id, client_secret = get_oidc_config()
 
     token_url, headers, body = client.prepare_token_request(
-        google["token_endpoint"],
-        authorization_response=f.request.url.replace("http://", "https://"),
-        redirect_url=f.request.base_url.replace("http://", "https://"),
+        oidc_provider_config["token_endpoint"],
+        authorization_response=f.request.url,  # .replace("http://", "https://"),
+        redirect_url=f.request.base_url,  # replace("http://", "https://"),
```

Comment:
Those schema rewrites were a way to work around the `InsecureTransportError` explained in a code comment above.

A tidier way to deal with this error is the `OAUTHLIB_INSECURE_TRANSPORT` environment variable, as also described in said code comment.

The rewrites, I believe, were only necessary when the current state was initially developed, i.e. with Conbench running locally, using Google's OIDC provider.

Reply:
_nods_ cool, we should be able to remove them totally then, yeah?

Reply:
Yes! Will do in one of the next PRs.

Reply:
Done in https://github.com/conbench/conbench/pull/457.